Source/Core/InputCommon/ControllerInterface/SDL/SDL.cpp

```diff
@@ -321,38 +329,129 @@ static constexpr SDLMotionAxisList SDL_AXES_GYRO = {{
 }};
 // clang-format on
 
-Joystick::Joystick(SDL_Joystick* const joystick, const int sdl_index)
-    : m_joystick(joystick), m_name(StripWhitespace(GetJoystickName(sdl_index)))
-{
-  if (SDL_JoystickNumButtons(joystick) > 255 || SDL_JoystickNumAxes(joystick) > 255 ||
-      SDL_JoystickNumHats(joystick) > 255 || SDL_JoystickNumBalls(joystick) > 255)
+GameController::GameController(SDL_GameController* const gamecontroller,
+                               SDL_Joystick* const joystick, const int sdl_index)
+    : m_gamecontroller(gamecontroller), m_sdl_index(sdl_index), m_joystick(joystick)
+{
+  const char* name;
+  if (gamecontroller)
+    name = SDL_GameControllerName(gamecontroller);
+  else
+    name = SDL_JoystickName(joystick);
+  m_name = name != NULL ? name : "Unknown";
+
+  // If a Joystick Button has a GameController equivalent, don't detect it
+  int n_legacy_buttons = SDL_JoystickNumButtons(joystick);
+  int n_legacy_axes = SDL_JoystickNumAxes(joystick);
+  int n_legacy_hats = SDL_JoystickNumHats(joystick);
+  std::vector<bool> is_button_mapped(n_legacy_buttons);
+  std::vector<bool> is_axis_mapped(n_legacy_axes);
+  std::vector<bool> is_hat_mapped(n_legacy_hats);
+
+  auto RegisterMapping = [&](SDL_GameControllerButtonBind bind) {
+    switch (bind.bindType)
+    {
+    case SDL_CONTROLLER_BINDTYPE_NONE:
+      return;
+    case SDL_CONTROLLER_BINDTYPE_BUTTON:
+      is_button_mapped.assign(bind.value.button, true);
+      break;
+    case SDL_CONTROLLER_BINDTYPE_AXIS:
+      is_axis_mapped.assign(bind.value.axis, true);
+      break;
+    case SDL_CONTROLLER_BINDTYPE_HAT:
+      is_hat_mapped.assign(bind.value.hat.hat, true);
+      break;
+    }
+  };
+
+  if (gamecontroller != nullptr)
   {
-    // This device is invalid, don't use it
-    // Some crazy devices(HP webcam 2100) end up as HID devices
-    // SDL tries parsing these as joysticks
-    return;
+    // Inputs
+
+    // Buttons
+    for (u8 i = 0; i != size(s_sdl_button_names); ++i)
+    {
+      SDL_GameControllerButton button = static_cast<SDL_GameControllerButton>(i);
+      if (SDL_GameControllerHasButton(m_gamecontroller, button))
+      {
+        AddInput(new Button(gamecontroller, button));
+        RegisterMapping(SDL_GameControllerGetBindForButton(gamecontroller, button));
+      }
+    }
+
+    // Axes
+    for (u8 i = 0; i != size(s_sdl_axis_names); ++i)
+    {
+      SDL_GameControllerAxis axis = static_cast<SDL_GameControllerAxis>(i);
+      if (SDL_GameControllerHasAxis(m_gamecontroller, axis))
+      {
+        // First 4 axes are for the analog sticks, the rest are for the triggers
+        if (i < 4)
+        {
+          // Each axis gets a negative and a positive input instance associated with it
+          AddAnalogInputs(new Axis(m_gamecontroller, -32768, axis),
```

Comment:
It seems like `AddAnalogInputs` isn't being used correctly here, which isn't difficult because the naming could be better.

`AddAnalogInputs` is specifically for handling inputs that have an unknown neutral position. An example is DirectInput or "legacy" SDL axes which might represent an analog stick with a neutral position of 0 or potentially a trigger with a neutral position of e.g. -32768. `AddAnalogInputs` adds additional "Full Surface" inputs to handle the case of the trigger that uses the full axis value range.

If we know the type of input and the neutral position for certain, which seems to be the case here with the "GameController" interface, just `AddInput` can be used. Adding the "Full Surface" inputs for a analog stick is unnecessary. It's only needed when something might be a trigger that uses the full negative-to-positive range.